### PR TITLE
POC - Enable js defined theme to be used in SASS

### PR DESCRIPTION
**Pull request by dprokop**
TODO: 
- [x] Allow JS variables in SASS files
- [x] Update `ThemeProvider` to expose active theme, instead of theme name
- [x] Update theme variable usage in @grafana/ui components to remove dependency on `variables.scss`
- [x] Update color picker styles to use theme
- [x] Fix stories for colorpickers

## Core changes
This PR introduces JS as a primary source of theme variables for Grafana. Theme definitions are located in `@grafana/ui/themes` directory. 

#### Themes are implemented in pure js. 
This is because our goal is to share variables between app and SASS. To achieve that themes are necessary during build time to be exposed to sass loader via `node-sass` functions (https://github.com/sass/node-sass/blob/master/README.md#functions--v300---experimental). This retrieval is implemented in `getThemeVariable(variablePath, themeName)`.

#### Themes are available to React components via `ThemeContext`
Context is available via `import  { ThemeContext } from '@grafana/ui';`

**If you skip `themeName` param, then dark theme's variant will be used**

### SASS
`getThemeVariable` is a function, that's available in sass files. Use it i.e. like this:
```scss
// In theme agnostic SASS file
.foo {
  font-size: getThemeVariable("typography.size.m");
}

// In *.[themeName].scss
.bar {
  background-color: getThemeVariable("colors.blueLight", "[themeName]")
}
```

#### Using `ThemeContext` directly
```ts
import  { ThemeContext } from '@grafana/ui';

<ThemeContext.Consumer>{
  (theme) => <Foo theme={theme} />
}</ThemeContext.Consumer>
```

#### Using `withTheme` HOC
With this method your component will be automatically wrapped in `ThemeContext.Consumer` and provided with current theme via `theme` prop. Component used with `withTheme` must implement `Themeable` interface. 
```ts
import  { ThemeContext, Themeable } from '@grafana/ui';

interface FooProps extends Themeable {}

const Foo: React.FunctionComponent<FooProps> = () => ...

export default withTheme(Foo);
```

### Storybook
All stories are wrapped with `ThemeContext.Provider` using global decorator. To render `Themeable` component that's not wrapped by `withTheme` HOC you either create a new component in your story:
```tsx
// Foo.story.tsx
const FooWithTheme = withTheme(Foo);

FooStories.add('Story' () => {
  return <FooWithTheme />
});
```
or use `renderComponentWithTheme` helper:

```tsx
// Bar.story.tsx

BarStories.add('Story' () => {
  return renderComponentWithTheme(Bar, /* pass props here */)
});
```
### Angular
There should be very few cases when theme would be used in Angular context. For this purpise there is a function available that retrieves current theme: `import { getCurrentTheme } from app/core/utils/ConfigProvider`

## Limitations
- #### Hot updates
Changes in JS theme files *are not subject of hot updates* during development. This applies to styles that comes from SASS files (which means 100% until we introduce css in js approach). This is a consequence of the fact that `getThemeVariable` util is executed during webpack pipeline. 
- #### You must ensure ThemeContext provider is available in a React tree
By default all react2angular directives have `ThemeContext.Provider` ensured. But, there are cases where we create another React tree via `ReactDOM.render`. This happens in case of graph legend rendering and `ReactContainer` directive. In such cases theme consumption will fail. To make sure theme context is available in such cases, you need to wrap your rendered component with ThemeContext.Provider using `provideTheme` function:

```typescript
// graph.ts
import { provideTheme } from 'app/core/utils/ConfigProvider';

// Create component with ThemeContext.Provider first.
// Otherwise React will create new components every time it renders!
const LegendWithThemeProvider = provideTheme(Legend);

const legendReactElem = React.createElement(LegendWithThemeProvider, legendProps);    
ReactDOM.render(legendReactElem, this.legendElem, () => this.renderPanel());
```
`provideTheme` makes current theme available via ThemeContext by checking if user has `lightTheme` set in her boot data.

### CODE REVIEW

**Comment by dprokop** [comment]
@torkelo @peterholmberg @Ijin08 @hugohaggmark @jschill 

Wanted to ask you guys for an opinion. The `GrafanaTheme` type is limited ATM as this work is kind of POC. I see in our variable.scss files that there are quite a lot of component specific variables defined there, e.g. `$input-color`. Values for those variables differs significantly across themes meaning that above variable can be `$gray-4` in dark theme and `$green` in light theme.

If we want to us css-in-js approach this means that we:
1. Introduce component specific variables into the themes like we have in sass vars
or 
2. We end up with a lot of code like this in our component's styles: 
```ts 
const selectedSwatchBorder = theme.type === GrafanaThemeType.Light ? '#ffffff' : '#1A1B1F';
```

Do you have any thoughts about that?

**Comment by torkelo** [comment]
> `$input-color`. Values for those variables differ significantly across themes meaning that above variable can be `$gray-4` in dark theme and `$green` in light theme.

yes, that is why we have the variables.dark & light. It proved impossible to have 2 palettes with the same names that can be used in the same way in light & dark. Originally I wanted the colors & grays to be same in light & dark but it did not work really well. Maybe with a different palette it could work. 

> 1. Introduce component specific variables into the themes like we have in sass vars

I would like to avoid that and move component variables into the component style file. 

> 2. const selectedSwatchBorder = theme.type === GrafanaThemeType.Light ? '#ffffff' : '#1A1B1F';

That looks horrible. Need to avoid that any way we can. One way would be to design a new palette with colors that can be used the same way in both light & dark. So they would be sort of inversed (the 
gray/dark scale colors at least). 

The other option would be
```ts
 const selectedSwatchBorder = selectColor({dark: theme.gray1, light: theme.dark2}); 
```

or something like that

**Comment by hugohaggmark** [comment]
Awesome work! I've understand the limitations and they seem reasonable.

**Comment by jschill** [comment]
> > 1. Introduce component specific variables into the themes like we have in sass vars
> 
> I would like to avoid that and move component variables into the component style file.

Agree. The rule of thumb could be to move the colors to the theme when two components use the same style. Ie `OurInputComponent` and `OurTextareaComponent` use the same border color so we create `FormElementBorderColor` in the theme-file.


> ```ts
>  const selectedSwatchBorder = selectColor({dark: theme.gray1, light: theme.dark2}); 
> ```

It would scale well. 
 ```ts
  const selectedSwatchBorder = selectColor({
    dark: theme.gray1,
    light: theme.dark2,
    custom: theme.blue,
    demoForManagement: theme.hotpink
}); 
 ```

**Comment by torkelo** [comment]
Did some research and looking how Kibana does some of this. The seem to be using styled-components for some parts. But some plugins seem to only support one theme as they import one theme. 


https://github.com/elastic/kibana/blob/8604e8f3f4e034adaa1fabff9c5e43a289f54284/x-pack/plugins/apm/public/components/shared/KueryBar/Typeahead/Suggestion.js#L19

// interesting function mixin 
https://github.com/elastic/eui/blob/ea17ac5400e038487495f6f48251b7957a343f52/src/global_styling/functions/_colors.scss#L32

https://github.com/elastic/kibana/blob/90554c174478ca0f9eb1735a79d8f6ccaeddc5ea/src/ui/public/vis/map/_leaflet_overrides.scss#L35

**Comment by torkelo** [comment]
IT strange in some components they take in theme via props and in others they just import one theme:

https://github.com/elastic/kibana/blob/8604e8f3f4e034adaa1fabff9c5e43a289f54284/x-pack/plugins/apm/public/components/shared/KueryBar/Typeahead/Suggestion.js#L19

File: public/sass/_variables.dark.scss [hunk 1]
```diff
@@ -129,7 +132,7 @@ $list-item-shadow: $card-shadow;
 $empty-list-cta-bg: $gray-blue;
 
 // Scrollbars
-$scrollbarBackground: #404357;
+$scrollbarBackground: #aeb5df;
```
Comment (torkelo) [inline public/sass/_variables.dark.scss]: accidental change?

Comment (dprokop) [reply public/sass/_variables.dark.scss]: Yes

Comment (torkelo) [reply public/sass/_variables.dark.scss]: noticed that the scrollbars looked different in the dark theme

Comment (dprokop) [reply public/sass/_variables.dark.scss]: I have noticed that yestarday as well and forgot to change. My bad.

File: jest.config.js [hunk 2]
```diff
@@ -6,11 +6,13 @@ module.exports = {
   },
   "moduleDirectories": ["node_modules", "public"],
   "roots": [
+    "<rootDir>/scripts",
     "<rootDir>/public/app",
     "<rootDir>/public/test",
     "<rootDir>/packages"
   ],
   "testRegex": "(\\.|/)(test)\\.(jsx?|tsx?)$",
+  "testPathIgnorePatterns": ["webpack.test.js"],
```
Comment (torkelo) [inline jest.config.js]: think we can delete webpack.test.js cannot find this being used anywhere

**Pull request merged**